Libraries/NativeAnimation/RCTNativeAnimatedModule.m

```diff
@@ -162,32 +218,47 @@ - (void)setBridge:(RCTBridge *)bridge
                   eventName:(nonnull NSString *)eventName
                   eventMapping:(NSDictionary<NSString *, id> *)eventMapping)
 {
-  [_operations addObject:^(RCTNativeAnimatedNodesManager *nodesManager) {
+  [self addOperationBlock:^(RCTNativeAnimatedNodesManager *nodesManager) {
     [nodesManager addAnimatedEventToView:viewTag eventName:eventName eventMapping:eventMapping];
   }];
 }
 
 RCT_EXPORT_METHOD(removeAnimatedEventFromView:(nonnull NSNumber *)viewTag
                   eventName:(nonnull NSString *)eventName)
 {
-  [_operations addObject:^(RCTNativeAnimatedNodesManager *nodesManager) {
+  [self addOperationBlock:^(RCTNativeAnimatedNodesManager *nodesManager) {
     [nodesManager removeAnimatedEventFromView:viewTag eventName:eventName];
   }];
 }
 
 #pragma mark -- Batch handling
 
+- (void)addOperationBlock:(AnimatedOperation)operation
+{
+  [_operations addObject:operation];
+}
+
+- (void)addPreOperationBlock:(AnimatedOperation)operation
+{
+  [_preOperations addObject:operation];
+}
+
 - (void)batchDidComplete
 {
-  NSArray *operations = _operations;
+  [_operationsLock lock];
+  NSArray<AnimatedOperation> *preOperations = _preOperations;
+  _preOperations = [NSMutableArray new];
+  NSArray<AnimatedOperation> *operations = _operations;
   _operations = [NSMutableArray new];
+  [_operationsLock unlock];
+
+  OperationBatch *batch =
+    [[OperationBatch alloc] initWithUIManager:self.bridge.uiManager
+                                 nodesManager:_nodesManager
+                                preOperations:preOperations
+                                   operations:operations];
 
-  dispatch_async(dispatch_get_main_queue(), ^{
-    [operations enumerateObjectsUsingBlock:^(AnimatedOperation operation, NSUInteger i, BOOL *stop) {
-      operation(self->_nodesManager);
-    }];
-    [self->_nodesManager updateAnimations];
-  });
+  [self.bridge.uiManager addUIManagerObserver:batch];
```

Comment:
Good call about set ordering, didnt think about that, I'll try the approach you suggested.